buildDotnetModule: allow fetching dependencies at compile time (#162548)
NixOS/nixpkgs: doc/languages-frameworks/dotnet.section.md

## Local Development Workflow {#local-development-workflow}

> For local development, its recommended to use nix-shell to create a dotnet environment:

Reviewer: ```suggestion
For local development, it's recommended to use `nix-shell` to create a dotnet environment:
```

}
```

> This will produce a dotnet installation that has the dotnet 3.1, 3.0, and 2.1 sdk. The first sdk listed will have its cli utility present in the resulting environment. Example info output:

Reviewer: ```suggestion
This will produce a dotnet installation that has the dotnet 3.1, 3.0, and 2.1 SDK. The first SDK listed will have its CLI utility present in the resulting environment. Example info output:
```

To package Dotnet applications, you can use `buildDotnetModule`. This has similar arguments to `stdenv.mkDerivation`, with the following additions:

* `projectFile` has to be used for specifying the dotnet project file relative to the source root. These usually have `.sln` or `.csproj` file extensions. This can be a list of multiple projects as well.

> * `nugetSha256` is used to specify the hash of the generated nix-compatible lockfile containing all Nuget dependencies. This depends on no floating version ranges being requested upstream, as we cannot deterministically fetch those. If any unstable version range is requested, an error will be thrown and you will need to use `nugetDeps` instead. This option is mutually exclusive with the `nugetDeps` attribute.

Reviewer: ```suggestion
* `nugetSha256` is used to specify the hash of the generated Nix-compatible lockfile containing all NuGet dependencies.

  This requires that no floating version ranges are requested upstream, as we cannot deterministically fetch those. If any unstable version range is requested, an error will be thrown and you will need to use `nugetDeps` instead. This option is mutually exclusive with the `nugetDeps` attribute.
```

> * `nugetDeps` is used to specify a manually generated nix-compatible lockfile, in case we cannot deterministically do so. A script to generate this file is available as `passthru.fetch-deps`. It can also be generated manually using `nuget-to-nix` tool, which is available in nixpkgs. If the argument is a derivation, it will be used directly and assume it has the same output as `mkNugetDeps`. This option is mutually exclusive with the `nugetSha256` attribute.

Reviewer: ```suggestion
* `nugetDeps` is used to specify a manually generated Nix-compatible lockfile, in case we cannot deterministically do so.

  A script to generate this file is available as `passthru.fetch-deps`. It can also be generated manually using `nuget-to-nix` tool, which is available in Nixpkgs. If the argument is a derivation, it will be used directly and assume it has the same output as `mkNugetDeps`. This option is mutually exclusive with the `nugetSha256` attribute.
```

* `dotnetPackFlags` can be used to pass flags to `dotnet pack`. Used only if `packNupkg` is set to `true`.
* `dotnetFlags` can be used to pass flags to all of the above phases.

> When packaging a new application, you need to fetch and lock it's Nuget dependencies. `buildDotnetModule` has two ways of handling this:

Reviewer 2: `it's` -> `its`

* Using `nugetSha256`. We try to generate a nix-compatible lockfile at build time, of which you can specify the hash with `nugetSha256`. When you try to build the derivation, everything gets fetched automatically and the expected hash gets printed out. We try this by default when neither `nugetSha256` nor `nugetDeps` are set, and it is the recommanded way to resolve dependencies. The downside of this method is that it only works if no floating versions are requested upstream, in which case an error will be thrown and you will need to use `nugetDeps` instead.

> * Using `nugetDeps`. This method requires you to manually generate a nix-compatible lockfile, which has to be checked into source control. You can run `nix-build -A package.fetch-deps` to build a script which generates the lockfile. After running said script you should copy the generated lockfile to a stable directory, and can set the `nugetDeps` attribute to it's path. Note that if you change any flags effecting the restore process, you will need to regenerate this script. This takes more work to maintain, so you should use `nugetSha256` instead when possible.

Reviewer 2: `it's` -> `its`

Note that both of these methods use the flags and attributes from your derivation, and attempt to mirror the restore process from the package itself. That means that if you for example set `dotnetRestoreFlags`, it will be used both for fetching dependencies and for building the derivation.

If you change the `dotnet-sdk` attribute, or set a flag effecting `dotnet restore`, you will most likely need to regenerate the lockfile.

> When updating an existing application, you should *always* regenerate it's lockfile.

Reviewer 2: `it's` -> `its`

* `dotnetPackFlags` can be used to pass flags to `dotnet pack`. Used only if `packNupkg` is set to `true`.
* `dotnetFlags` can be used to pass flags to all of the above phases.

When packaging a new application, you need to fetch and lock its Nuget dependencies. `buildDotnetModule` has two ways of handling this:

> * Using `nugetSha256`. We try to generate a nix-compatible lockfile at build time, of which you can specify the hash with `nugetSha256`. When you try to build the derivation, everything gets fetched automatically and the expected hash gets printed out. We try this by default when neither `nugetSha256` nor `nugetDeps` are set, and it is the recommanded way to resolve dependencies. The passthru attribute `nuget-lockfile` is exposed which builds this lockfile for you. The downside of this method is that it only works if no floating versions are requested upstream, in which case an error will be thrown and you will need to use `nugetDeps` instead.

Reviewer: I don't understand what this means:
> passthru attribute `nuget-lockfile` is exposed 

Does it mean there exists an attribute `passthru.nuget-lockfile` to the argument of `buildDotnetModule`? This should be made explicit.

```suggestion
* Using `nugetSha256`. We try to generate a Nix-compatible lockfile at build time, of which you can specify the hash with `nugetSha256`. When you try to build the derivation, everything gets fetched automatically and the expected hash gets printed out. We try this by default when neither `nugetSha256` nor `nugetDeps` are set. This is the recommended way to resolve dependencies. The passthru attribute `nuget-lockfile` is exposed which builds this lockfile for you. The downside of this method is that it only works if no floating versions are requested upstream – otherwise an error will be thrown and you will need to use `nugetDeps` instead.
```

> * Using `nugetDeps`. This method requires you to manually generate a nix-compatible lockfile, which has to be checked into source control. You can run `nix-build -A package.fetch-deps` to build a script which generates the lockfile. After running said script you should copy the generated lockfile to a stable directory, and can set the `nugetDeps` attribute to its path. Note that if you change any flags effecting the restore process, you will need to regenerate this script. This takes more work to maintain, so you should use `nugetSha256` instead when possible.

Reviewer: ```suggestion
* Using `nugetDeps`. This method requires you to manually generate a Nix-compatible lockfile, which has to be checked into source control. You can run `nix-build -A package.fetch-deps` to build a script which generates the lockfile. After running said script you can set the `nugetDeps` attribute to the path of the generated lockfile. Note that if you change any flags affecting the restore process, you will need to regenerate this script. This takes more work to maintain, so you should use `nugetSha256` instead when possible.
```

The wording was unclear, and I think it's unnecessary to remind readers that the generated file should have a predictable path relative to the Nix expression.
